Updated _check_shape for BaseClassification

anmolsjoshi opened the pull request:
Fixes #555  

Description:
Added check to squeeze last dimension if equals to 1. 

Check list:
* [x] New tests are added (if a new feature is added)
* [ ] New doc strings: description and/or example code are in RST format
* [ ] Documentation is updated (if required)

vfdev-5 commented:
@anmolsjoshi I thought about to remove this squeeze operations. Actually, I'm not sure about their need...

anmolsjoshi commented:
@vfdev-5 were you thinking of doing shape check like this - y.squeeze().shape == y_pred.squeeze().shape ?

vfdev-5 commented:
@anmolsjoshi I was thinking to remove these lines :

https://github.com/pytorch/ignite/blob/f50a3f31336dcf16a128f0d1917d6ed28b296a46/ignite/metrics/accuracy.py#L24-L30

anmolsjoshi commented:
@vfdev-5 processing it like that, allows for the [_check_type](https://github.com/pytorch/ignite/blob/f50a3f31336dcf16a128f0d1917d6ed28b296a46/ignite/metrics/accuracy.py#L51) function to work correctly. Allowing the distinguishing between binary and multiclass cases.

vfdev-5 commented:
OK, that's good to know. 
Anyway, the logic to remove a dimension with 1 can either generalized to remove all dimensions with 1 or either keep it as is and update our further logic.
Imagine another case, where user has  somehow `y_pred` as `(B, C, 1, S)` and `y` as `(B, 1, S)`, current code will fail again...

vfdev-5 commented on ignite/metrics/accuracy.py:
```diff
@@ -20,14 +20,8 @@ def reset(self):
-            # (N, 1, ...) -> (N, ...)
-            y_pred = y_pred.squeeze(dim=1)
+        y_pred = y_pred.squeeze()
+        y = y.squeeze()
```
nice :)

anmolsjoshi replied:
@vfdev-5 I just realized this fails in case of batch size = 1.

I think we should squeeze all dimensions except for the 1st. Thoughts?

vfdev-5 replied:
make sense, we can try like that

anmolsjoshi replied:
@vfdev-5 have a look at the updated code, let me know your thoughts.

vfdev-5 commented on ignite/metrics/accuracy.py:
```diff
@@ -18,16 +18,17 @@ def reset(self):
         self._type = None
         self._num_classes = None
 
+    def _process_shape(self, y):
```
This method is ugly, sorry

vfdev-5 replied:
I would prefer not to squeeze at all and adapt our logic and raise an error if we can not understand the case.

anmolsjoshi replied:
I updated the function. 

What logic would you use to check the shapes?

vfdev-5 replied:
I'm trying to analyse here. 
As documentation states:
```
    - `y_pred` must be in the following shape (batch_size, num_categories, ...) or (batch_size, ...).
    - `y` must be in the following shape (batch_size, ...).
```
we must have batch size of `y` and batch size of `y_pred` to be equal and all dimensions hided behind `...` to be equal for `y` and `y_pred`.

This logic should be sufficient ?

Next, we need to determine the case : multi-class or (binary/multi-label). I think `y.ndimension() + 1 == y_pred.ndimension()` should hold even if we do not squeeze 1's.

anmolsjoshi replied:
@vfdev-5 I mostly agree. Here's a situation.

Let's take an example of a binary case of batch size N - model output shape is (N, 1) and training set shape is (N, ). Given logic of check_type, this case will be considered a multiclass case, because y_pred.ndimension() == y.ndimension() + 1.

Should we have users account for situations using output_transform? I personally think we should account for this situation in the metrics. Most updated way of handling 1's in the shape, I think works well.

vfdev-5 replied:
Or we can assign a multi-class case when second dimension is larger than one ?

anmolsjoshi replied:
Clarifying, if a user inputs (N, 10) and (N, 1) for multiclass cases. That is considered wrong by the new logic right? We should raise an error for this?

vfdev-5 replied:
We can follow this decision "tree" (https://github.com/pytorch/ignite/pull/275#issuecomment-425396475) and update it to the current state.

anmolsjoshi commented:
@vfdev-5 please have a look at the updated branch. I changed the metrics, tests per your comment.

anmolsjoshi commented:
@vfdev-5 we currently don't have restrictions on that, do you want to throw an error or have the metric handle it?

vfdev-5 commented:
That's fine without restrictions. I want just to make sure that such corner cases are well handled.

anmolsjoshi commented:
I updated the logic for check_type to handle this case, and updated how binary accuracy was calculated to prevent any errors. 

These are the errors I wanted to avoid:
1. _type is recorded as multiclass
2. the binary checks are not implemented.

vfdev-5 commented on ignite/metrics/accuracy.py:
```diff
@@ -122,12 +121,12 @@ def reset(self):
 
         if self._type == "binary":
-            correct = torch.eq(y_pred.type(y.type()), y).view(-1)
+            correct = torch.eq(y_pred.view(-1).type(y.type()), y.view(-1))
```
Maybe we can change the following
```
y_pred.view(-1).type(y.type())
```
by 
```
y_pred.view(-1).to(y)
```
https://pytorch.org/docs/stable/tensors.html#torch.Tensor.to

anmolsjoshi replied:
agreed and updated.

vfdev-5 commented:
Joshi, can you please add the tests from the original issue such that the PR fixes the issue too

anmolsjoshi commented:
Victor, I added 1 example and another test for the new check_type logic.

Did you want me to add to tests (described in the issue) to all other test cases and other metrics?

vfdev-5 commented:
I thinks it is OK for now. We can merge this once travis is green
